src/cargo/core/resolver/mod.rs

```diff
@@ -1098,3 +1087,51 @@ fn check_duplicate_pkgs_in_lockfile(resolve: &Resolve) -> CargoResult<()> {
     }
     Ok(())
 }
+
+/// re-run all used resolve_features so it can print warnings
+fn emit_warnings(
+    cx: &Context,
+    resolve: &Resolve,
+    summaries: &[(Summary, Method<'_>)],
+    config: Option<&Config>,
+) {
+    let mut new_cx = cx.clone();
+    new_cx.resolve_features = im_rc::HashMap::new();
+    let mut features_from_dep = HashMap::new();
+    for (summery, method) in summaries {
+        for (dep, features) in new_cx
+            .resolve_features(None, summery, &method, config)
+            .expect("can not resolve_features for a required summery")
+        {
+            features_from_dep.insert((summery.package_id(), dep), features);
+        }
+    }
+    for summery in resolve.sort().iter().rev().map(|id| {
+        cx.activations
+            .get(&id.as_activations_key())
+            .expect("id in dependency graph but not in activations")
+            .0
+            .clone()
+    }) {
+        for (parent, deps) in cx.parents.edges(&summery.package_id()) {
+            for dep in deps.iter() {
+                let features = features_from_dep
+                    .remove(&(*parent, dep.clone()))
+                    .expect("fulfilled a dep that was not needed");
+                let method = Method::Required {
+                    dev_deps: false,
+                    features: &features,
+                    all_features: false,
+                    uses_default_features: dep.uses_default_features(),
+                };
+                for (dep, features) in new_cx
+                    .resolve_features(None, &summery, &method, config)
```

Comment:
Would it be possible to perhaps avoid re-calling `resolve_features` here? I found this somewhat complicated to follow, and figured it might be simpler if we simply walk over the graph after resolution, check the list of activated features for each package, and then present a warning if the feature enables a required dependency.

We may produce a slightly worse error message since we don't have *why* a feature is activated, but that seems reasonable given the age of this warning perhaps?

Reply:
The original commit https://github.com/rust-lang/cargo/pull/6860/commits/f0f856502eb5b27e9c5766d86b0a95bfd1f98fc6, did that. It was a lot cleaner. But it did not work. `resolve_features` is only adds a canonicalized feature to `cx.resolve_features`. So I did not find a way to reconstruct the warnings from that. (The hard cases are "invalid9", "dep_feature_in_cmd_line") Suggestions are welcome.

I also don't think we can extend that strategy to only the features transitively enabled by some deps. But I would love to be proven wrong.

Reply:
I wonder if we could build up a side table or something like that during resolution to facilitate this strategy of emitting warnings? I suppose that's sort of like what happens today, but I think it'd be best to decouple the warning emission as much as we can from the exact state of the resolver today, because it may make future changes to `resolve_features` more complicated in the future

Reply:
True, this new code that redos the work of the resolver is going to be a pane to keep in sink.

I gave up on making it work, and the future opportunities of filtering dependencies, and just made this case a resolver bactrackable error.